### PR TITLE
ci: use trusted publishing for release

### PR DESCRIPTION
**Pull request by bluetech**
See https://docs.pypi.org/trusted-publishers/. Loosely based on pytest CI.

### CODE REVIEW

**Comment by bluetech** [comment]
We'll see if it works on release, might need fixups...

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+
+      - name: Build and Check Package
+        uses: hynek/build-and-inspect-python-package@2dbbf2b252d3a3c7cec7a810e3ed5983bd17b13a  # v2.8.0
+
+  deploy:
+    if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags') && github.repository == 'pytest-dev/pytest-django'
```
Comment (webknjaz) [inline .github/workflows/deploy.yml:25]: FYI, there's a `github.ref_name` available now.

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    steps:
+      - name: Download Package
+        uses: actions/download-artifact@v4
+        with:
+          name: Packages
+          path: dist
+
+      - name: Publish package
+        uses: pypa/gh-action-pypi-publish@8a08d616893759ef8e1aa1f2785787c0b97e20d6 # v1.10.0
```
Comment (webknjaz) [inline .github/workflows/deploy.yml:43]: With this version, it's possible to opt in to uploading digital attestations, FYI.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -126,34 +124,3 @@ jobs:
...
-        with:
-          python-version: '3.12'
-
-      - name: Install dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install --upgrade build
-
-      - name: Build package
-        run: python -m build
```
Comment (webknjaz) [inline .github/workflows/main.yml:153]: It's usually nice to keep it in the main workflow and put the job building the artifacts in front of the tests. Then, it'd be possible to test whatever's going to be uploaded, as opposed to building something new and potentially differing.

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (bluetech) [reply .github/workflows/deploy.yml:25]: Not sure about `github.ref_name` but maybe `github.ref_type` is better for this?

File: .github/workflows/deploy.yml (hunk 2 above)
Comment (bluetech) [reply .github/workflows/deploy.yml:43]: Will look into it for the next release

File: .github/workflows/main.yml (hunk 3 above)
Comment (bluetech) [reply .github/workflows/main.yml:153]: Agree, but I didn't want to complicate/slow down the test CI for now, but wanted to at least do it right for the deploy workflow.

**Pull request merged**